### ISSUE TITLE
EmailAddressValidator does not work with IDN domains

### ISSUE THREAD

**Issue by lorenzulrich** [neos/flow-development-collection#1227]
### Description

An e-mail address using an IDN domain name is not considered a valid e-mail address.

### Steps to Reproduce

1. Create a scenario using an EmailAddressValidator, e.g. an e-mail field with `Neos.Form`.
2. Enter an e-mail address with an IDN domain, e.g. `test@neös.io`.
3. Trigger validation

#### Expected behavior

The e-mail address is valid.

#### Actual behavior

A validation error is thrown: `Please specify a valid email address.`

### Affected Versions

Flow: all versions

### Background

The EmailAddressValidator is using `filter_var` as follows:

```filter_var($emailAddress, FILTER_VALIDATE_EMAIL)```

This does not support IDN domain names. As of PHP 7.1, it seems to be possible to set a filter flag `FILTER_FLAG_EMAIL_UNICODE`, but this is not a short-term solution for Flow installations using PHP < 7.1.

**Comment by Torsten85** [comment]
What about replacing all non printable characters? This would change `test@neös.io` to `test@nes.io` which would be a valid email address. Non valid email addresses would still fail the test.

`$emailAddress = preg_replace( '/[^[:print:]]/', '',$emailAddress);`

**Comment by dlubitz** [comment]
Sounds a bit hacky. Why not convert/encode them?
An lib for example: https://github.com/true/php-punycode

**Comment by lorenzulrich** [comment]
I agree that it's too hacky. Imagine an e-mail address `test@äää.com` which would be reduced to `test@.com`, making it invalid. Not a very probable use case in German, but if we're thinking e.g. russian letters only, it might be different.
Converting however seems to be a good idea.

**Comment by sorenmalling** [comment]
```
php > var_dump(filter_var('soren@' . idn_to_ascii(end(explode('@', 'soren@täst.de'))), FILTER_VALIDATE_EMAIL) !== false);
bool(true)
php >
```

Modified the current email address validator for a test purpose - seems legit. Is this a backport for earlier versions as well ?

**Comment by albe** [comment]
I would prefer to fix this for Flow 5.x+ with the `FILTER_FLAG_EMAIL_UNICODE` flag and then see what can be done about lower versions. Unfortunately, a first quick test with that flag showed, that it will not work with IDN domains, just with unicode characters in the address part.

The issue with `idn_to_ascii` is, that it requires the php_intl extension to be enabled, which it isn't by default in all installations. So given it's ok to add this requirement, I'd probably go with just adding that to the validator. However I'm not sure it is really acceptable to expect an additional extension in old versions to have email validation still work, since that would be breaking. So it would need an check for the function to be available and then, that would suddenly become a switch in behaviour depending on the enabled extensions. I don't like such hard to follow behaviour.

So with all that in mind, my suggestion would be to add the `idn_to_ascii` feature to the upcoming 5.1 release earliest.

**Comment by albe** [comment]
I'd like to tackle this for the upcoming release with adding a dependency to `ext-intl` and using `idn_to_ascii` in the validator. This should be relatively straightforward, so anyone wanting to take a stab at this is welcome!

Also, I'd like to somehow enhance the EmailValidator even further if possible, because currently it does a very tiny part of what is actually useful for validating e-mail addresses. Symfony for example allows checking against different patterns (especially the one used for HTML5 form fields is probably useful) as well as validate the DNS records (MX, or A/AAAA/MX) - and this is very important when going to send e-mails to those adresses to reduce bounces.
See https://symfony.com/doc/current/reference/constraints/Email.html
However, this goes beyond the scope of this issue itself, so I just mention it here for the record.
